chore: tsoa dev watch

IrakliJani opened the pull request:
Closes: <!-- reference the related issue e.g. #150 -->

### Description:

upgrades tsoa to latest and adds dev watch for tsoa

- `yarn workspace backend dev-tsoa`

this is an experiment and should be merged to `yarn dev` if it succeeds

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

IrakliJani commented on packages/backend/src/controllers/catalogController.ts:
```diff
@@ -46,7 +46,7 @@ export class CatalogController extends BaseController {
      * - search: string
      * - type: 'table' | 'field'
-     * @returns
+     * @returns ApiCatalogResults
```
returns without value now throws an error with the new tsoa version

IrakliJani commented on packages/backend/package.json:
```diff
@@ -112,8 +114,9 @@
     },
     "scripts": {
         "generate-api": "tsoa spec-and-routes --configuration tsoa.yml",
+        "generate-api-dev": "chokidar './src/controllers/**/*.ts' -c 'yarn run generate-api'",
```
im not going to introduce this as part of the dev mode (for now)

IrakliJani replied:
let's see if it's effective and then introduce it to the dev

rephus replied:
we should also regenerate the api if any underlying model type changes, which is harder to do...

IrakliJani commented on packages/backend/src/database/entities/savedSemanticViewerCharts.ts:
```diff
@@ -71,9 +71,9 @@ export type DBSavedSemanticViewerChartVersion = {
     saved_semantic_viewer_chart_uuid: string;
     created_at: Date;
-    config: object;
+    config: Record<string, unknown>;
```
inferring objects or {} when generating swagger causes warnings. I just followed instructions to get rid of all warnings

rephus replied:
should a type from backend entities be exposed on the API ? we normally don't expose these. 

And also, I remember tsoa was not able to handle Record types, it was simply removing their values, which was hard to debug.

IrakliJani commented on packages/common/src/types/dbt.ts:
```diff
@@ -452,22 +452,22 @@ export const convertModelMetric = ({
         dimensionReference,
         requiredAttributes,
-        ...(metric.urls ? { urls: metric.urls } : {}),
+        ...(metric.urls ? { urls: metric.urls } : null),
```
same as {}/object what I mentioned above. inferring {} is weird with ts

rephus commented on packages/backend/src/generated/swagger.json:
I guess this new swagger should be compatible with our lightdash docs. We could copy this new file into the lightdash-docs repo and make a PR to see if docusaurus is compatible 

let me know if you want me to do it

rephus commented on packages/common/src/types/filter.test.ts:
```diff
@@ -19,7 +20,7 @@ describe('compress and uncompress dashboard filters', () => {
                 tableName: 'payments',
             },
-            tileTargets: {},
+            tileTargets: {} as Record<string, DashboardTileTarget>,
```
is this change in the test necessary for tsoa ? This type hasn't changed, right ?

rephus reviewed (approved):
code looks good, I made a few comments, but they are not blocking

IrakliJani commented:
@rephus beautiful! thanks for testing that. im checking the breaking changes

```
error [request-property-all-of-added]
in API POST /api/v1/projects/{projectUuid}/dashboards
added '#/components/schemas/FilterRule_ConditionalOperator.DashboardFieldTarget.any.any_' to the '/anyOf[#/components/schemas/CreateDashboard]/filters/dimensions/items/' request property 'allOf' list
 
error [request-property-all-of-added]
in API POST /api/v1/projects/{projectUuid}/dashboards
added '#/components/schemas/FilterRule_ConditionalOperator.DashboardFieldTarget.any.any_' to the '/anyOf[#/components/schemas/CreateDashboard]/filters/metrics/items/' request property 'allOf' list
 
error [request-property-all-of-added]
in API POST /api/v1/projects/{projectUuid}/dashboards
added '#/components/schemas/FilterRule_ConditionalOperator.DashboardFieldTarget.any.any_' to the '/anyOf[#/components/schemas/CreateDashboard]/filters/tableCalculations/items/' request property 'allOf' list
 
error [response-property-became-optional]
in API DELETE /api/v1/projects/{projectUuid}/dataCatalog/{catalogSearchUuid}/categories/{tagUuid}
the response property 'results' became optional for the status '200'
 
error [response-property-became-optional]
in API GET /api/v1/projects/{projectUuid}/explores
the response property 'results/items/anyOf[subschema #2]/allOf[#/components/schemas/Pick_ExploreError.SummaryExploreErrorFields_]/tags' became optional for the status '200'
 
error [response-property-type-changed]
in API GET /api/v1/projects/{projectUuid}/user-credentials
the 'results/credentials/anyOf[#/components/schemas/Pick_CreateRedshiftCredentials-or-CreatePostgresCredentials-or-CreateSnowflakeCredentials-or-CreateTrinoCredentials.type-or-user_]/type' response's property type/format changed from 'string'/'' to ''/'' for status '200'
 
error [response-property-type-changed]
in API GET /api/v1/user/warehouseCredentials
the 'results/items/credentials/anyOf[#/components/schemas/Pick_CreateRedshiftCredentials-or-CreatePostgresCredentials-or-CreateSnowflakeCredentials-or-CreateTrinoCredentials.type-or-user_]/type' response's property type/format changed from 'string'/'' to ''/'' for status '200'
 
error [response-property-type-changed]
in API POST /api/v1/user/warehouseCredentials
the 'results/credentials/anyOf[#/components/schemas/Pick_CreateRedshiftCredentials-or-CreatePostgresCredentials-or-CreateSnowflakeCredentials-or-CreateTrinoCredentials.type-or-user_]/type' response's property type/format changed from 'string'/'' to ''/'' for status '200'
 
error [response-property-type-changed]
in API PATCH /api/v1/user/warehouseCredentials/{uuid}
the 'results/credentials/anyOf[#/components/schemas/Pick_CreateRedshiftCredentials-or-CreatePostgresCredentials-or-CreateSnowflakeCredentials-or-CreateTrinoCredentials.type-or-user_]/type' response's property type/format changed from 'string'/'' to ''/'' for status '200'
 
warning [request-property-all-of-removed]
in API POST /api/v1/projects/{projectUuid}/dashboards
removed '#/components/schemas/FilterRule_ConditionalOperator.T.V.any_' from the '/anyOf[#/components/schemas/CreateDashboard]/filters/dimensions/items/' request property 'allOf' list
 
warning [request-property-all-of-removed]
in API POST /api/v1/projects/{projectUuid}/dashboards
removed '#/components/schemas/FilterRule_ConditionalOperator.T.V.any_' from the '/anyOf[#/components/schemas/CreateDashboard]/filters/metrics/items/' request property 'allOf' list
 
warning [request-property-all-of-removed]
in API POST /api/v1/projects/{projectUuid}/dashboards
removed '#/components/schemas/FilterRule_ConditionalOperator.T.V.any_' from the '/anyOf[#/components/schemas/CreateDashboard]/
```

this is the list and just trying to figure out the impact from each
